storage: unblock the GC queue for large transaction cleanups

spencerkimball opened the pull request:
Transactions with enough intent spans to GC have proven problematic when
the intents cannot be resolved within the timeout. This can lead to the
GC queue timing out without ever being able to finish a range.

This PR removes a temporary fix to limit synchronous intent cleanups to
a maximum of 800 and replaces that with two different initiatives:

- Expired transaction records are now processed entirely asynchronously
  using the intent resolver's `processIntentsAsync()`. This avoids hanging
  normal processing of GC'able spans.
 - `intentResolver.resolveIntents` was previously creating additional
   async tasks for every batch, which while the batching alleviates the
   quadratic slowdown from noop entries in the batches, would end up with
   up to 100 batches in parallel, each with 100 intents. That would be a
   massive hit to any system, much less one with 3 nodes. Seems better to
   simplify this method and run the batches serially, each with its own
   timeout. Intent resolution is inherently amortizable; there's no need
   for this kind of rush.
 - `intentResolver.resolveIntents` now sets a separate timeout for every
   batch of 100 intents. The idea here is that we want to always be making
   forward progress within the intent resolution timeout, as opposed to
   being able to finish resolving every intent within the timeout.
- GC'able spans are chunked and GC'ed *before* we move to pushing txns
  and resolving intents for user keyspace spans. This prevents us from
  doing all the work to scan the range only to never send the GC request
  because we timeout trying to push potentially many txns and resolve
  potentially many intents.

spencerkimball committed:
storage: unblock the GC queue for large transaction cleanups

Transactions with enough intent spans to GC have proven problematic when
the intents cannot be resolved within the timeout. This can lead to the
GC queue timing out without ever being able to finish a range.

This PR removes a temporary fix to limit synchronous intent cleanups to
a maximum of 800 and replaces that with two different initiatives:

- Expired transaction records are now processed entirely asynchronously
  using the intent resolver's `processIntentsAsync()`. This avoids hanging
  normal processing of GC'able spans.
 - `intentResolver.resolveIntents` was previously creating additional
   async tasks for every batch, which while the batching alleviates the
   quadratic slowdown from noop entries in the batches, would end up with
   up to 100 batches in parallel, each with 100 intents. That would be a
   massive hit to any system, much less one with 3 nodes. Seems better to
   simplify this method and run the batches serially, each with its own
   timeout. Intent resolution is inherently amortizable; there's no need
   for this kind of rush.
 - `intentResolver.resolveIntents` now sets a separate timeout for every
   batch of 100 intents. The idea here is that we want to always be making
   forward progress within the intent resolution timeout, as opposed to
   being able to finish resolving every intent within the timeout.
- GC'able spans are chunked and GC'ed *before* we move to pushing txns
  and resolving intents for user keyspace spans. This prevents us from
  doing all the work to scan the range only to never send the GC request
  because we timeout trying to push potentially many txns and resolve
  potentially many intents.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20230)
<!-- Reviewable:end -->

spencerkimball committed:
storage: further improvements to GC queue processing

Reorganize code in gc queue so that the machinery already in
`intentResolver` is used to batch `PushTxn` requests for expired
transactions.

In `intentResolver`, chunk push txn requests into batches with a
maximum size. Also, split up code previously in `processIntents` to
two methods: `cleanupIntents` and `cleanupTxnIntents`. The latter
aims to cleanup the transaction record after resolving all intents
belonging to it.

tschottdorf commented:
Sorry to let this sit!

This (and by that I mean the first commit, which is all I've looked at so far) is good stuff! I'd like you to pull out the intent resolver concurrency changes into their own PR, as proposed in my inline comment below. But more importantly, I think you should extract and submit the change that GCs the keys before it goes on the fat-txn-record bonanza separately, as we'll be able to cherry-pick that and that alone should get us pretty close to removing `maxIntentHack`. The remaining part (reworking how we call intent resolution) is fantastic too, but I don't think we can justify it for 1.1.x, and so that too should be alone in a PR (but I think it's just what's left over after my previous suggestions).

The second commit looks like more good cleanup. I think it's natural to squash it into the leftovers of the first one after the other changes have been factored out, though again, the part in which you refactor `processIntents` would better be pulled out again into an individual PR if you ask me.

I know that that's not pleasant work and that the current PR is a natural consequence of pulling on a thread, but this is a splendid opportunity to shine up a regularly neglected corner of our code that has given us a lot of trouble, which is ultimately that's what we're after here, and I'll be able to go after the details better when the diffs are less overwhelming.

---

Reviewed 3 of 4 files at r1, 1 of 1 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 384 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DX9LH7czIuglsvWWY:-L-DX9LH7czIuglsvWWZ:bqmqbcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L384)):*
> ```Go
> 	handleTxnIntents := func(key roachpb.Key, txn *roachpb.Transaction) error {
> 		if len(txn.Intents) > 0 {
> 			// Synthesize an EndTransaction request in order to send IntentsWithArgs.
> ```

Is there a specific reason you chose to go this route as opposed to always returning the GCKey plus sending the intents to the intent resolver manually? I think that would be preferable as now you're shoehorning something into an existing part of the system that assumes that each such input is a client committing the txn themselves (for example, it's not inconceivable that we'd put a metric in that path). I would feel more comfortable here if you called `resolveIntents` directly.

There is also the problem here that you're firing off an unbounded number of goroutines (one per txn), though after the discussed `intent_resolver` changes those would get forced through a semaphore in `DistSender`. Regardless, I'd feel safer if you added a (generous) semaphore here (500?).

---

*[pkg/storage/gc_queue.go, line 615 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-D_-zA6VEL6rlGxurn:-L-D_-zA6VEL6rlGxuro:biiyolq) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L615)):*
> ```Go
> 			return nil
> 		},
> 		func(now hlc.Timestamp, txn *roachpb.Transaction, typ roachpb.PushTxnType) {
> ```

All of these methods should take a `ctx`, but I prefer if you just put a TODO here because we need to keep the diff small. (Make it out in my name if you like).

---

*[pkg/storage/gc_queue.go, line 628 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DQVSgVKHjUh0DdnPM:-L-DQVSgVKHjUh0DdnPN:b-f18ghi) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L628)):*
> ```Go
> 	}
> 
> 	log.Eventf(ctx, "MVCC stats: %+v", repl.GetMVCCStats())
> ```

Clarify what stats they are. They are now the stats *after* GC whereas previously it was before:

`log.Eventf(ctx, "MVCC stats after GC: %+v", repl.GetMVCCStats())`

---

*[pkg/storage/gc_queue.go, line 629 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DQkyhyCGMzU4Iqrzz:-L-DQkyhyCGMzU4Iqs--:bwkebx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L629)):*
> ```Go
> 
> 	log.Eventf(ctx, "MVCC stats: %+v", repl.GetMVCCStats())
> 	log.Eventf(ctx, "done GC'ing, new score is %s", makeGCQueueScore(ctx, repl, repl.store.Clock().Now(), sysCfg))
> ```

Just change this message to `GC Score after GC: %s` while you're here.

---

*[pkg/storage/gc_queue.go, line 696 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DQ6FeA7pVGMLaHeTy:-L-DQ6FeA7pVGMLaHeTz:b-5o4dbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L696)):*
> ```Go
> // RunGC runs garbage collection for the specified descriptor on the
> // provided Engine (which is not mutated). It uses the provided
> // pushTxnFn to clarify the true status of a transaction,
> ```

Lacks description of `gcFn` and in particular the pattern it will be invoked with (once?).

---

*[pkg/storage/gc_queue.go, line 866 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DZAwoY6VtZCFJ4-Z9:-L-DZAwoY6VtZCFJ4-ZA:bei478i) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L866)):*
> ```Go
> 
> 	// Clean up the AbortSpan.
> 	log.Event(ctx, "processing AbortSpan")
> ```

This part of the change is very good as it breaks [the infinite loop](https://github.com/cockroachdb/cockroach/issues/19216#issuecomment-339419528) by making sure the GC score reduces even when you have a fat txn record. In fact, it's so good that I'd like to get it into 1.1.x, which means that you'd have to extract it into its own PR that we merge first. That is hopefully straightforward, mind doing so?

---

*[pkg/storage/intent_resolver.go, line 304 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DWZ0rwpFFSjGrhNr2:-L-DWZ0rwpFFSjGrhNr3:b-pehzap) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L304)):*
> ```Go
> 			} else {
> 				return errors.Wrap(err, "intents not asynchonously processed as there "+
> 					"is no available async processing bandwidth")
> ```

If you don't have to check what error you have on your hands in this one, why is it checked in the other one?

---

*[pkg/storage/intent_resolver.go, line 420 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DRO59trnD4NAT0ovY:-L-DRO59trnD4NAT0ovZ:b-34c1jj) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L420)):*
> ```Go
> 	// TODO(bdarnell): Note that this functionality has been removed and
> 	// will be ignored, pending resolution of #8360.
> 	Wait   bool
> ```

Remove this field now that the underlying functionality has been removed.

---

*[pkg/storage/intent_resolver.go, line 426 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DT_8WN19sq16Jiud3:-L-DT_8WN19sq16Jiud4:bry7owg) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L426)):*
Could you pull the changes in this method into their own PR? I'm glad you're cleaning up this area, but any mistake here is detrimental to far more than GC, and I think we need to do some work testing this together.

> intentResolver.resolveIntents` was previously creating additional
   async tasks for every batch, which while the batching alleviates the
   quadratic slowdown from noop entries in the batches, would end up with
   up to 100 batches in parallel, each with 100 intents. That would be a
   massive hit to any system, much less one with 3 nodes. Seems better to
   simplify this method and run the batches serially, each with its own
   timeout. Intent resolution is inherently amortizable; there's no need
   for this kind of rush.

These changes are :100: but I think it should go a step further. Namely, that we actually have [fixed the quadratic memory growth](https://github.com/cockroachdb/cockroach/pull/16946) in `DistSender` and so the most efficient thing is to pass all of the requests to it at once, as it is internally best-equipped to parallelize the work that needs to be done (which it does). That is, let's take your code and remove the loop, and *that* should just perform well (yay!)

Then, we need to find out some way to stress test this, which is fortunately not that hard but a little tedious, but at least could motivate me to take https://github.com/cockroachdb/cockroach/pull/18661 somewhere, which is a prerequisite to feeling safe about changes here and in the GC queue (something we've long lacked).

---

*[pkg/storage/intent_resolver.go, line 488 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DS2oNmiTeHXvSW-p9:-L-DS2oNmiTeHXvSW-pA:b-61ps7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L488)):*
> ```Go
> 		cancel()
> 		if err != nil {
> 			// Bail out on the first error.
> ```

I was expecting this kind of code:

```go
- cancel()
if err != nil && ctxWithTimeout.Err() != nil && ctx.Err() == nil {
  log.Warningf(ctx, "skipping intent resolution after error %v", err)
}
+ cancel()
```

to allow a large batch to continue making forward progress even if individual chunks time out. What do you think?

---

*[pkg/storage/intent_resolver.go, line 395 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-Dc4I7wTuppY443B1X:-L-Dc4I7wTuppY443B1Y:b1hg032) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a4cb7f8fd3f83147008421e1c15789965735c9f/pkg/storage/intent_resolver.go#L395)):*
> ```Go
> 	}
> 	if updateGCMetrics {
> 		ir.store.metrics.GCResolveSuccess.Inc(int64(len(resolveIntents)))
> ```

Instead of having a random boolean here, I think it's preferable to have methods a la `cleanupIntentsOnGC` which do this part and then simple call their non-GC counterpart, as that is self-documenting, better to grep for, and harder to get wrong.

---

*[pkg/storage/intent_resolver.go, line 438 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DbMmBmfsSvZyM1otN:-L-DbMmBmfsSvZyM1otO:b3tfv4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a4cb7f8fd3f83147008421e1c15789965735c9f/pkg/storage/intent_resolver.go#L438)):*
> ```Go
> 	ctx context.Context,
> 	txn *roachpb.Transaction,
> 	intents []roachpb.Intent,
> ```

The intent here is flawed because you're passing a transaction but the intents each also contain one. I see that you're blindly overwriting them below, but 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-:-L-D_EuMZqvMfyGLC2H9:b45inz0)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
OK, I'm pushing an update with changes to things which stopped short of splitting this into multiple separate PRs. But your statement: "I think you should extract and submit the change that GCs the keys before it goes on the fat-txn-record bonanza separately" is confusing with regard to this change because:

- This change sends extant txn records (fat and slim ones alike) for asynchronous resolution.
- It still sends those for async processing *first*, not after GC'ing the keys.
- This is necessary because we can't very well GC the txn record until *after* we've resolved all of its intents.

My proposal for splitting is as follows:

PR 1 (for cherry-pick to 1.1.x):
 - Minimal changes to `gc_queue.go` and `intent_resolver.go` to send `COMMITTED` and `ABORTED` txns for async processing and to then perform GC on keys *before* moving to resolve intents encountered during the range scan. This is the lightest footprint change I can envision which makes sure that we GC keys and also cleanup fat txns any time a range is processed by the GC queue.

PR 2 (for 1.2):
- Remove batching from `intentResolver`. This is actually a bit of an iffy change and shouldn't go into 1.1.x because it's going to change the dynamic for how intense the load can become when resolving fat txns.

PR 3 (for 1.2):
- Refactor `intentResolver` to have a wider API, suitable for use from `gc_queue.go`, refactor `gc_queue.go` to use the expanded API, including removing the existing, ugly code in the queue which manually pushes txns with a semaphore and loose goroutines, as well as the current failure to both push and the resolve intents for a `PENDING` but expired txn.

---

Review status: 3 of 5 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/storage/gc_queue.go, line 384 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DX9LH7czIuglsvWWY:-L-FbO2HgU2iuFg6d9nA:bgfak1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L384)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Is there a specific reason you chose to go this route as opposed to always returning the GCKey plus sending the intents to the intent resolver manually? I think that would be preferable as now you're shoehorning something into an existing part of the system that assumes that each such input is a client committing the txn themselves (for example, it's not inconceivable that we'd put a metric in that path). I would feel more comfortable here if you called `resolveIntents` directly.

There is also the problem here that you're firing off an unbounded number of goroutines (one per txn), though after the discussed `intent_resolver` changes those would get forced through a semaphore in `DistSender`. Regardless, I'd feel safer if you added a (generous) semaphore here (500?).
</blockquote></details>

In the second commit I made the interface to the `intentResolver` cleaner, so we're no longer shoehorning this into the `intentResolver` API which expects an `IntentsWithArg` parameter. The problem with calling `resolveIntents` directly is that that code is not asynchronous, and it feels like the task of cleaning up potentially large txns and then removing them (can only do this after cleanup of all intents) shouldn't be in the critical path of the GC.

This code isn't firing off an unbounded number of goroutines. It actually relies on the limited async task manager used in `intentResolver.cleanupTxnIntentsAsync`. The limit on that is 500, and it's shared with the rest of the system's intent resolution efforts, which seems like a good thing. And of course you're correct that there's also the `DistSender` limits in the event that the intents here end up being on many different ranges.

My ultimate goal with this change was to remove all of the duplicative intent and txn handling code and centralize everything around the intent resolver. That seems like a worthy goal. This stuff deviated mostly because of Ben and I working separately at different times.

---

*[pkg/storage/gc_queue.go, line 615 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-D_-zA6VEL6rlGxurn:-L-Fd-mFXCA9Pp2ORtQg:b-aj0wiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L615)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

All of these methods should take a `ctx`, but I prefer if you just put a TODO here because we need to keep the diff small. (Make it out in my name if you like).
</blockquote></details>

I will add contexts to these methods when I figure out how to separate this into multiple changes for sequencing. The contexts can get added last so they don't mess with the earlier efforts to build a fix for 1.1.

---

*[pkg/storage/gc_queue.go, line 628 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DQVSgVKHjUh0DdnPM:-L-FdME5bytrbiGM3d2A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L628)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Clarify what stats they are. They are now the stats *after* GC whereas previously it was before:

`log.Eventf(ctx, "MVCC stats after GC: %+v", repl.GetMVCCStats())`
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 629 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DQkyhyCGMzU4Iqrzz:-L-FdQwq-OFa3IPNgkcX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L629)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Just change this message to `GC Score after GC: %s` while you're here.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 696 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DQ6FeA7pVGMLaHeTy:-L-Fdav6uDGzVbtxKoYB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L696)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Lacks description of `gcFn` and in particular the pattern it will be invoked with (once?).
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 866 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DZAwoY6VtZCFJ4-Z9:-L-FdfCWFhv9JitMw6r7:b-t76d4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L866)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This part of the change is very good as it breaks [the infinite loop](https://github.com/cockroachdb/cockroach/issues/19216#issuecomment-339419528) by making sure the GC score reduces even when you have a fat txn record. In fact, it's so good that I'd like to get it into 1.1.x, which means that you'd have to extract it into its own PR that we merge first. That is hopefully straightforward, mind doing so?
</blockquote></details>

Yes, I can break this part out. Don't we also need to take the big txns out of the critical path as part of a PR to cherry pick into 1.1.x?

---

*[pkg/storage/intent_resolver.go, line 304 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DWZ0rwpFFSjGrhNr2:-L-FeT9QmsTOudUdrf5i:bwzh5ci) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L304)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

If you don't have to check what error you have on your hands in this one, why is it checked in the other one?
</blockquote></details>

I don't understand what you're asking here. Could you elaborate?

---

*[pkg/storage/intent_resolver.go, line 420 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DRO59trnD4NAT0ovY:-L-FhPcPSwWZDAXYcw7A:b6y0s5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L420)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Remove this field now that the underlying functionality has been removed.
</blockquote></details>

My issue with that is this `Wait` is a very useful performance optimization and removing the intention of whether we ought to wait or not from all of the call sites isn't just a lot of code churn, it's also going to be harder to be convinced we're not getting something wrong when we actually fix #8360 and re-enable this. I feel more comfortable leaving the `Wait` option alone and just removing the unused implementation with this TODO in place.

---

*[pkg/storage/intent_resolver.go, line 426 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DT_8WN19sq16Jiud3:-L-FjAzHC0g6_d2GwRUU:b-z7i7uj) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L426)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Could you pull the changes in this method into their own PR? I'm glad you're cleaning up this area, but any mistake here is detrimental to far more than GC, and I think we need to do some work testing this together.

> intentResolver.resolveIntents` was previously creating additional
   async tasks for every batch, which while the batching alleviates the
   quadratic slowdown from noop entries in the batches, would end up with
   up to 100 batches in parallel, each with 100 intents. That would be a
   massive hit to any system, much less one with 3 nodes. Seems better to
   simplify this method and run the batches serially, each with its own
   timeout. Intent resolution is inherently amortizable; there's no need
   for this kind of rush.

These changes are :100: but I think it should go a step further. Namely, that we actually have [fixed the quadratic memory growth](https://github.com/cockroachdb/cockroach/pull/16946) in `DistSender` and so the most efficient thing is to pass all of the requests to it at once, as it is internally best-equipped to parallelize the work that needs to be done (which it does). That is, let's take your code and remove the loop, and *that* should just perform well (yay!)

Then, we need to find out some way to stress test this, which is fortunately not that hard but a little tedious, but at least could motivate me to take https://github.com/cockroachdb/cockroach/pull/18661 somewhere, which is a prerequisite to feeling safe about changes here and in the GC queue (something we've long lacked).
</blockquote></details>

OK, I've dramatically simplified this. Let's hope that the full parallelization doesn't crush things, but I agree better to keep it simple. Pulling things out into new changes is the next step across the board for this PR. As mentioned in the top-level comment, best that this only goes into 1.2.

---

*[pkg/storage/intent_resolver.go, line 488 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DS2oNmiTeHXvSW-p9:-L-FjtxyqcePK0jsIrp-:brbnrde) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L488)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I was expecting this kind of code:

```go
- cancel()
if err != nil && ctxWithTimeout.Err() != nil && ctx.Err() == nil {
  log.Warningf(ctx, "skipping intent resolution after error %v", err)
}
+ cancel()
```

to allow a large batch to continue making forward progress even if individual chunks time out. What do you think?
</blockquote></details>

I think this suggestion is moot now that we're no longer batching.

---

*[pkg/storage/intent_resolver.go, line 395 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-Dc4I7wTuppY443B1X:-L-FgUUAnmuvZqI4iMAo:bx6ns96) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a4cb7f8fd3f83147008421e1c15789965735c9f/pkg/storage/intent_resolver.go#L395)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Instead of having a random boolean here, I think it's preferable to have methods a la `cleanupIntentsOnGC` which do this part and then simple call their non-GC counterpart, as that is self-documenting, better to grep for, and harder to get wrong.
</blockquote></details>

OK, rearranged things so this boolean is gone.

---

*[pkg/storage/intent_resolver.go, line 438 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DbMmBmfsSvZyM1otN:-L-Fh42-mNp92DwU9zFx:b-m2k301) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a4cb7f8fd3f83147008421e1c15789965735c9f/pkg/storage/intent_resolver.go#L438)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The intent here is flawed because you're passing a transaction but the intents each also contain one. I see that you're blindly overwriting them below, but 
</blockquote></details>

They're all intents which belong to the txn. I could verify that...the code before was looking at the first intent to get the txn status and txn meta. That seemed uglier IMO. We could also drop the intents and instead collect spans and pass those here instead so there's less confusion. Other ideas?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-:-L-FjzJ6sg2AMEbXAg41:b-81xtq2)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> But your statement: "I think you should extract and submit the change that GCs the keys before it goes on the fat-txn-record bonanza separately" is confusing with regard to this change because:

Yeah, I agree, that change alone needs more complications to be addressed before it is useful. Your suggested plan looks good, looking forward to PR#1! I can dust off my WIP PRs that generate lots of these fat txns to see whether 1.1.x with that patch can chew through them.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 384 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DX9LH7czIuglsvWWY:-L-I2Es1ftvdNMHpYlUl:b-c3nezz) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L384)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

In the second commit I made the interface to the `intentResolver` cleaner, so we're no longer shoehorning this into the `intentResolver` API which expects an `IntentsWithArg` parameter. The problem with calling `resolveIntents` directly is that that code is not asynchronous, and it feels like the task of cleaning up potentially large txns and then removing them (can only do this after cleanup of all intents) shouldn't be in the critical path of the GC.

This code isn't firing off an unbounded number of goroutines. It actually relies on the limited async task manager used in `intentResolver.cleanupTxnIntentsAsync`. The limit on that is 500, and it's shared with the rest of the system's intent resolution efforts, which seems like a good thing. And of course you're correct that there's also the `DistSender` limits in the event that the intents here end up being on many different ranges.

My ultimate goal with this change was to remove all of the duplicative intent and txn handling code and centralize everything around the intent resolver. That seems like a worthy goal. This stuff deviated mostly because of Ben and I working separately at different times.
</blockquote></details>

Sorry, I missed the upshot of what you're doing, which is not having to tie the removal of the txn record into the GC queue's process timeout at all. Pretty happy about the combined diff I'm seeing here.

---

*[pkg/storage/gc_queue.go, line 866 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DZAwoY6VtZCFJ4-Z9:-L-I0CO7LtfF8cl4LV_4:b-e6o8qt) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L866)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Yes, I can break this part out. Don't we also need to take the big txns out of the critical path as part of a PR to cherry pick into 1.1.x?
</blockquote></details>

The reason the big txns are such a problem right now is because they prevent the keys from ever being GC'ed. With this change, you'd only DDOS yourself whenever the range becomes eligible for GC again (the txn itself gives zero priority). But yeah, ideally we can solve the big txn problem in general. I just couldn't estimate whether the resulting diff would be too dangerous for `1.1.x`.

---

*[pkg/storage/intent_resolver.go, line 304 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DWZ0rwpFFSjGrhNr2:-L-I0dqwLUet7KGvGahG:ble6taw) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L304)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I don't understand what you're asking here. Could you elaborate?
</blockquote></details>

You take the other branch only when `err == stop.ErrThrottled`, but in this branch you're also assuming `err == stop.ErrThrottled` but it's not necessarily true (except by knowing that that's the only thing `err` will ever be if not nil).

---

*[pkg/storage/intent_resolver.go, line 420 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DRO59trnD4NAT0ovY:-L-I1287mcwEwwUvjBMq:bg2j5hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L420)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

My issue with that is this `Wait` is a very useful performance optimization and removing the intention of whether we ought to wait or not from all of the call sites isn't just a lot of code churn, it's also going to be harder to be convinced we're not getting something wrong when we actually fix #8360 and re-enable this. I feel more comfortable leaving the `Wait` option alone and just removing the unused implementation with this TODO in place.
</blockquote></details>

Makes sense, thanks.

---

*[pkg/storage/intent_resolver.go, line 438 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DbMmBmfsSvZyM1otN:-L-I0wOTzhBPLe9yctAl:b-jceema) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a4cb7f8fd3f83147008421e1c15789965735c9f/pkg/storage/intent_resolver.go#L438)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

They're all intents which belong to the txn. I could verify that...the code before was looking at the first intent to get the txn status and txn meta. That seemed uglier IMO. We could also drop the intents and instead collect spans and pass those here instead so there's less confusion. Other ideas?
</blockquote></details>

It's a bit clunky, but I'm fine leaving as is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-:-L-I1rZnfd2bW8NQh61w:b8gjhj3)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 384 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DX9LH7czIuglsvWWY:-L-IHNs0IFAhIwb7HW4S:b-89q77x) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L384)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Sorry, I missed the upshot of what you're doing, which is not having to tie the removal of the txn record into the GC queue's process timeout at all. Pretty happy about the combined diff I'm seeing here.
</blockquote></details>

Cool!

---

*[pkg/storage/gc_queue.go, line 866 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DZAwoY6VtZCFJ4-Z9:-L-IHTlyHeMXY2oLvzAb:b-tpx6be) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/gc_queue.go#L866)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The reason the big txns are such a problem right now is because they prevent the keys from ever being GC'ed. With this change, you'd only DDOS yourself whenever the range becomes eligible for GC again (the txn itself gives zero priority). But yeah, ideally we can solve the big txn problem in general. I just couldn't estimate whether the resulting diff would be too dangerous for `1.1.x`.
</blockquote></details>

Do you think it's safe enough?

---

*[pkg/storage/intent_resolver.go, line 304 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DWZ0rwpFFSjGrhNr2:-L-IHmNg9gzjuJe1V8nB:bjc73my) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L304)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

You take the other branch only when `err == stop.ErrThrottled`, but in this branch you're also assuming `err == stop.ErrThrottled` but it's not necessarily true (except by knowing that that's the only thing `err` will ever be if not nil).
</blockquote></details>

Ah, gotcha. Will make this change in #20396 

---

*[pkg/storage/intent_resolver.go, line 420 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DRO59trnD4NAT0ovY:-L-IHsG4LJpQ3L8uiO1M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/81e8a9e431ae26c4fcd22d2e204f4a49fc2baebb/pkg/storage/intent_resolver.go#L420)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Makes sense, thanks.
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 438 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230#-L-DbMmBmfsSvZyM1otN:-L-IHrL1kuUnpxc1Q1GG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a4cb7f8fd3f83147008421e1c15789965735c9f/pkg/storage/intent_resolver.go#L438)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

It's a bit clunky, but I'm fine leaving as is.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20230)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Closed in favor of #20396, #20400, and #20404.
